lib/puppet/provider/nameservice.rb

```diff
@@ -276,6 +283,13 @@ def initialize(resource)
     super
     @custom_environment = {}
     @objectinfo = nil
+    if resource.is_a?(Hash) && !resource[:canonical_name].nil?
+      # we're instantiating from a hash, ie self.instances
+      @canonical_name = resource[:canonical_name]
```

Comment:
In this case we've queried the system Etc value to get a struct of user data (and subsequently it has been scrubbed of invalid data) and transformed that user data into a hash, which we're going to use to construct this provider instance.